src/twisted/conch/ssh/connection.py

```diff
@@ -65,8 +66,16 @@ def serviceStopped(self):
         """
         Called when the connection is stopped.
         """
-        for channel in list(self.channels.values()):
+        # Close any fully open channels
+        for channel in list(self.channelsToRemoteChannel.keys()):
```

Comment:
My understanding of the implicit contract that `channelClosed` is upholding with its test is that a channel that hasn't had either `channelOpen` or `openFailed` shouldn't receive a `closed` call either — the channel should be guaranteed to received either `channelOpen` or `openFailed` first, and to receive `closed` only after a `channelOpen`.

So in this case, the code I added in the next couple of lines is there to handle the channels which `channelClosed` refused to handle.

I could add an explanatory comment if you think it would be a good idea.